File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/PIDController.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.util.ElapsedTime;


public class PIDController {

    double Kp;
    double Ki;
    double Kd;
    double lastError = 0;
    double integral = 0;
    boolean angleWrap = false;

    ElapsedTime timer = new ElapsedTime();

    /**
     * Set PID gains
     * @param Kp proportional gain
     * @param Ki integral gain
     * @param Kd derivative gain
     */
    public PIDController(double Kp, double Ki, double Kd) {
        this.Kp = Kp;
        this.Ki = Ki;
        this.Kd = Kd;
    }

    public PIDController(double Kp, double Ki, double Kd, boolean angleWrap) {
        this.Kp = Kp;
        this.Ki = Ki;
        this.Kd = Kd;
        this.angleWrap = angleWrap;
    }

    /**
     * calculate PID output given the reference and the current system state
     * @param reference where we would like our system to be
     * @param state where our system is
     * @return the signal to send to our motor or other actuator
     */
    public double update(double reference, double state) {
        double error;
        double derivative;
        // check if we need to unwrap angle
        if (angleWrap) {
            error = angleWrap(reference - state);
        } else {
            error = reference - state;
        }
        // define deriative value
        derivative = (error - lastError) / timer.seconds();

        // forward euler integration
        integral += error * timer.seconds();
        if (integral > 1) {
            integral = 1;
        }
        else if (integral < -1) {
            integral = -1;
        }

        double output = (error * Kp) + (integral * Ki) + (derivative * Kd);

        timer.reset();
        lastError = error; 

        return output;
    }


    public double angleWrap(double radians) {
        while (radians > Math.PI) {
            radians -= 2 * Math.PI;
        }
        while (radians < -Math.PI) {
            radians += 2 * Math.PI;
        }
        return radians;
    }


}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Kotlin_Bromine_Arya/Auto/PID_Components/PIDController.kt
package org.firstinspires.ftc.teamcode.Kotlin_Bromine_Arya.Auto.PID_Components

import com.qualcomm.robotcore.util.ElapsedTime

//TODO(Move to A util class)


class PIDController {
    private val timer: ElapsedTime = ElapsedTime()
    private var kp: Double = 0.0
    private var ki: Double = 0.0
    private var kd: Double = 0.0
    private var kf: Double = 0.0
    private var integral: Double = 0.0
    private var beforeError: Double = 0.0
    private var derivitive: Double = 0.0
    private var error: Double = 0.0

    fun setPIDF(kp: Double, ki: Double, kd: Double, kf:Double = 0.0) {
        this.kp = kp
        this.ki = ki
        this.kd = kd
        this.kf = kf
    }

    fun calculate(current: Double, target: Double): Double {
        error = target - current

        integral += (error * timer.seconds())

        derivitive = (error - beforeError) / timer.seconds()

        beforeError = error

        timer.reset()

        return derivitive * kd + integral * ki + error * kp
    }
}